Add a script to generate configupdater vault entry for new clusters

alvaroaleman opened the pull request:
Ref https://issues.redhat.com/browse/DPTP-2413

/assign @hongkailiu

petr-muller commented on hack/generate_config_updater_kubeconfig.sh:
```diff
@@ -0,0 +1,27 @@
+  exit 1
+fi
+
+oc config use-context "$CLUSTER"
```
Can we use `--context` in individual calls to avoid side effects for whoever runs this?

hongkailiu replied:
This line would fail (and the whole script exit with failure) if the context $CLUSTER is not found.
Would it ensure no side-effects out of running this?

hongkailiu reviewed (commented):
Steve convinced me to run our existing tools locally for this card.
1. `applyconfig` to generate objects on the cluster.
2. `secret-generator` to upload secrets to Vault.

I admit the advantage:
We maintain only one set of tools. Although the script is quite simple, still there is possibility that we change those tools. We need to copy those changes over to this script.

So I would create a new Make target in this repo which runs the 2 tools above.
WDYT?

alvaroaleman commented:
> Steve convinced me to run our existing tools locally for this card.

Yeah, doesn't really make much difference one way or another. I've changed the script to a maketarget and fixed up the existing ci-scret-bootstrap maketarget to use it.

The maketarget uses `oc apply` because that does the same thing as applyconfig here but is simpler, as we don't have to mount everything into a container. PTAL.

hongkailiu commented:
/retest

hongkailiu commented on Makefile:
```diff
@@ -275,3 +268,9 @@ openshift-image-mirror-mappings:
+config_updater_vault_secret:
+	@[[ -z $$cluster ]] && echo "ERROR: \$$cluster must be set" && exit 1
+	oc apply --context=$(cluster) --recursive -f clusters/build-clusters/common
+	make ci-secret-bootstrap
```
Shouldnt this be `ci-secret-generator`?
the data have not been uploaded to Vault yet.

hongkailiu commented on Makefile:
```diff
@@ -210,21 +210,14 @@ job:
+	@./hack/ci-secret-bootstrap.sh
 .PHONY: ci-secret-bootstrap
 
 ci-secret-generator:
```
I will revisit this after my changes on the job.

hongkailiu replied:
Maybe add this target and use it as a pre-step of `ci-secret-generator`? The ugly thing like `- cmd: oc --kubeconfig /tmp/build-farm-credentials/sa.config-updater.$(cluster).config sa create-kubeconfig` need to be taken care of

```
build_farm_credentials_folder ?= /tmp/build-farm-credentials

build_farm_credentials_folder:
	mkdir -p $(build_farm_credentials_folder)
	oc --context app.ci -n ci extract secret/config-updater --to=$(build_farm_credentials_folder) --confirm
.PHONY: build_farm_credentials_folder
```

hongkailiu commented on Makefile:
```diff
@@ -275,3 +268,9 @@ openshift-image-mirror-mappings:
+
+config_updater_vault_secret:
+	@[[ -z $$cluster ]] && echo "ERROR: \$$cluster must be set" && exit 1
+	oc apply --context=$(cluster) --recursive -f clusters/build-clusters/common
```
Would it fail if the file for the namespace comes after the file for the objects in the namespace?
I guess that is one of the advantages of using `applyconfig`.

hongkailiu commented on Makefile:
```diff
@@ -275,3 +259,16 @@ openshift-image-mirror-mappings:
+		--config-dir=/manifests \
+		--context=$(cluster) \
+		--kubeconfig=/_kubeconfig
+	make hack/ci-secret-generator.sh
```
```suggestion
	make ci-secret-generator
```

hongkailiu replied:
We also need to land the kubeconfig for the new cluster after `applyconfig` and before `ci-secret-generator`

```
oc --context "$(cluster)" sa create-kubeconfig -n ci config-updater > "$(build_farm_credentials_folder)/sa.config-updater.$(cluster).config"
```

hongkailiu commented on hack/ci-secret-generator.sh:
```diff
@@ -0,0 +1,30 @@
+cluster="${cluster:-}"
+echo $BASE_DIR
+
+set -x
```
debugging leftover?
